---
title: "Differential Expression Analysis"
output:
  bookdown::html_document2: default
bibliography: differential_expression.bib
---


```{r}
#| label: setup
#| include: false

knitr::opts_chunk$set(
  echo = TRUE,
  include = TRUE
  )
```

# Linear Models

Approach underlying, e.g., limma [@Ritchie2015]. Formulated
initially in the context of analysis of microarray data,
e.g. [@Smyth2004]. For the $g^{\text{th}}$ gene, let
$y_{g1}, \dots, y_{gn}$ be the measured response values. For
microarray data, for example, $y_{gi}$ is usually either
the log-ratio $\log_2 R_{gi} - \log_2 G_{gi}$ (for two-colour
arrays) or log-intensity (for one colour arrays) measured
for gene $g$ in array $i$.

After normalising to remove unwanted technical variation, the
response is assumed to be described by the linear model
\begin{equation}
\mathrm{E}[\mathbf{y}_g] = X \mathbf{\alpha}_g ,
\end{equation}
where $\mathbf{y}_g^\prime = (y_{g1},\dots,y_{gn})$, $X$ is 
a suitable design matrix, and $\mathbf{\alpha}_g$ is the
coefficient vector. The variance is allowed to vary across
genes, such that
\begin{equation}
\mathrm{Var}[\mathbf{y}_g] = W_g \sigma_g^2 ,
\end{equation}
where $W_g$ is a (known) non-negative matrix of weights.

Biological contrasts of interest are encoded as
\begin{equation}
\mathbf{\beta}_g = C^\prime \mathbf{\alpha}_g ,
\end{equation}
for suitable $C$. The general problem in a differential
expression analysis is to assess whether particular contrasts
within this set are zero.

Estimates $\hat{\mathbf{\alpha}}_g$ and $s_g^2$ of the unknown parameters
$\mathbf{\alpha}_g$ and $\sigma_g^2$ are assumed to be available,
with estimated covariance matrices
\begin{equation}
\mathrm{Var}[\hat{\mathbf{\alpha}}_g] = V_g s_g^2 .
\end{equation}
In the case that $V_g$ depends on $\mathbf{\alpha}_g$, then it is
assumed that $V_g$ is evaluated at the estimate $\hat{\mathbf{\alpha}}_g$,
with this dependence ignorable to first-order. Estimators for the
contrasts of interest are
\begin{equation}
\hat{\mathbf{\beta}}_g = C^\prime \hat{\mathbf{\alpha}}_g ,
\end{equation}
with
\begin{equation}
\mathrm{Var}[\hat{\mathbf{\beta}}_g] = C^\prime V_g C s_g^2 .
\end{equation}
To develop the hierarchical model proposed in [@Smyth2004], it
is assumed that the contrasts $\hat{\mathbf{\beta}}_g$ are
normally distributed with mean $\mathbf{\beta}_g$ and
covariance matrix $C^\prime V_g C \sigma_g^2$, and $s_g^2$
is approximately distributed according to a scaled
$\chi^2$ distribution,
\begin{equation}
\begin{aligned}
s_g^2 | \sigma_g^2 &\sim \frac{\sigma_g^2}{d_g} \chi_{d_g}^2 , \\
\hat{\beta}_{gj} | \beta_{gj}, \sigma_g^2 &\sim
N(\beta_{gj}, v_{gj} \sigma_g^2) ,
\end{aligned}
\end{equation}
where $v_{gj}$ is the $j^{\text{th}}$ diagonal element of
$C^\prime V_g C$ and $d_g$ is the residual d.f. for the linear
model for gene $g$. This implies that 
$\hat{\beta}_{gj} / \sigma_{g} \sqrt{v_{gj}} |
\beta_{gj},\sigma_g^2 \sim N(\beta_{gj}, 1)$,
$s_g^2 d_g / \sigma_g^2 | \sigma_g^2 \sim \chi_{d_g}^2$,
and the test statistic
\begin{equation}
t_{gj} = \frac{\hat{\beta}_{gj}}{s_g \sqrt{v_{gj}}} =
\frac{\frac{\hat{\beta}_{gj}}{\sigma_g \sqrt{v_{gj}}}}
{\sqrt{\frac{s_g^2 d_g}{d_g \sigma_g^2}}}
\end{equation}
is approximately $t_{d_g}$-distributed under the null
$\beta_{gj} = 0$. Estimators for distinct genes $g_1$ and $g_2$
are assumed independent. For $\sigma_g^2$, a prior of the form
\begin{equation}
\frac{1}{\sigma_g^2} \sim \frac{1}{d_0 s_0^2} \chi_{d_0}^2
\end{equation}
is assumed. For $\beta_{gj}$, the prior is taken to be a mixture
of a point mass at zero, with mixing probability
$1 - \mathrm{Pr}(\beta_{gj} \neq 0) \equiv 1 - p_j$, and
the conjugate normal prior
\begin{equation}
\beta_{gj} | \sigma_g^2, \beta_{gj} \neq 0 \sim
N(0, v_{0j} \sigma_g^2) .
\end{equation}



```{r}
#| label: session-info
sessionInfo()
```
